include/deal.II/base/tensor_function_parser.h

```diff
@@ -264,17 +264,37 @@ class TensorFunctionParser : public TensorFunction<rank, dim, Number>
 private:
 #ifdef DEAL_II_WITH_MUPARSER
   /**
-   * Place for the variables for each thread
+   * std::is_copy_constructible gives the wrong answer for containers with
+   * non-copy constructible types (e.g., std::vector<std::unique_ptr<int>>).
+   * Work around that problem by defining our own explicitly uncopyable
+   * per-thread data class (instead of, e.g., using std::pair).
+   *
+   * We define exactly one per-thread data structure to avoid locking and
+   * unlocking the TLS data lookup more than once per function call.
    */
-  mutable Threads::ThreadLocalStorage<std::vector<double>> vars;
+  struct ParserData
+  {
+    ParserData() = default;
+
+    ParserData(const ParserData &) = delete;
+
+    /**
+     * Scratch array used to set independent variables (i.e., x, y, and z)
+     * before each muParser call.
+     */
+    std::vector<double> vars;
+
+    /**
+     * muParser implementations.
+     */
+    std::vector<std::unique_ptr<internal::muParserBase>> parsers;
+  };
```

Comment:
Isn't this the exact same class as for the `function_parser.h` file?

Reply:
Yes. I need to move some more functions into `mu_parser_internal.h` - if its OK with you I would like to do that refactoring in the next PR.

Reply:
Yes.